draft-sheffer-wimse-s2s-protocol.md

```diff
@@ -279,6 +279,109 @@ etc., are all valid and equivalent header field names. However, case is signific
 
 ## Option 1: DPoP-Inspired Authentication {#dpop-esque-auth}
 
+This option, inspired by the OAuth DPoP specification {{?RFC9449}}, uses a DPoP-like mechanism to authenticate
+the calling workload in the context of the request. The WIMSE Identity Token {{to-wit}} is sent in the request as
+described in {{wit-http-header}}. An additional JWT, the Workload Proof Token (WPT), is signed by the private key
+corresponding to the public key in the WIT. The WPT is sent in the `Workload-Proof-Token` header field of the request.
+A WPT contains the following:
+
+* in the JOSE header:
+    * `alg`: An identifier for an appropriate JWS asymmetric digital signature algorithm corresponding to
+     the confirmation key in the associated WIT.
+    * `typ`: the WPT is explicitly typed, as recommended in {{Section 3.11 of RFC8725}},
+     using the `application/wimse-proof+jwt` media type.
+* in the JWT claims:
+    * `iss`: The issuer of the token, which is the calling workload, represented by the same value as the `sub` claim
+     of the associated WIT.
+    * `aud`: The audience of the token contains the the HTTP target URI ({{Section 7.1 of RFC9110}}) of the request
+     to which the WPT is attached, without query or fragment parts.
+    * `iat`: The time at which the token was created (as defined in {{Section 4.1.6 of RFC7519}}).
+    * `jti`: A unique identifier for the token.
+    * `ctx_token_hash`: Hash of the OAuth access token, Transaction Token {{?I-D.ietf-oauth-transaction-tokens}},
+     or other token in the request that might convey end-user identity and authorization context of the request.
+     The value MUST be the result of a base64url encoding (as defined in {{Section 2 of RFC7515}}) the SHA-256 hash of
+     the ASCII encoding of the associated token's value.
+
```

Comment:
Can there be more than one context token hash?  would this need to be specified as an array or can you have multiple of the same claim in the jwt?

Reply:
> Can there be more than one context token hash? 

This is a good question that I asked myself while writing the above. But didn't have a great sense for so opted for the simple answer of just one. 

The whole construct is admittedly somewhat awkward - trying to account/allow for different kinds of tokens that might show up in different places. 

 
> would this need to be specified as an array or can you have multiple of the same claim in the jwt?

Multiple of the same claim in the jwt is not allowed (see https://datatracker.ietf.org/doc/html/rfc7519#section-4) so an array is how that kind of data is expressed.

Reply:
But maybe a few different claims that conditionally show up would be a better approach. Like `ath` access token hash - a la https://www.rfc-editor.org/rfc/rfc9449.html#section-4.2-6.2 - and then `txth` transaction token hash and maybe an oth for *other* token hash...

Reply:
Currently rethinking the construct

Reply:
Although we can definitely punt on it for now, I think we could choose a construct that's both simple and extensible.

```json
"signed_headers": {
    "txn-token": "txn token hash",
    "x-custom-context": "x-custom-context hash"
}
```

Unfortunately we need the WPT to sign miscellaneous stuff, because we don't want a new WIT for every call. Having said that, as far as I can tell it does NOT need to sign the WIT itself.

Reply:
https://github.com/yaronf/wimse-s2s/issues/25

Reply:
Created issue #25 from @yaronf's comment and c5b46a1d06829b5ec435f143d16d57a478003706 tries to do this in a somewhat better way.